### ISSUE TITLE
VirtualService applies rule to wrong Gateway

### ISSUE THREAD

**Issue by edussx** [istio/istio#10313]
**Describe the bug**
According to https://istio.io/docs/tasks/traffic-management/ingress/#configuring-ingress-using-an-istio-gateway, a VirtualService only applies its rules to the Gateways specified in the `gateways` field; however, my VirtualService applies to all the Gateways in the same namespace. For example, I created the following 3 Gateways(listen port 80, 81 and 82):
```
---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: case-d-gateway-1
  namespace: chenxma
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
        number: 80
        name: http
        protocol: HTTP
    hosts:
    - "*"
---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: case-d-gateway-2
  namespace: chenxma
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
        number: 81
        name: http
        protocol: HTTP
    hosts:
    - "*"
---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: case-d-gateway-3
  namespace: chenxma
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
        number: 82
        name: http
        protocol: HTTP
    hosts:
    - "*"
---
```
and 3 VirtualServices apply to the above Gateways respectively:
```
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: case-d-virtualservice-1
  namespace: chenxma
spec:
  hosts:
  - "*"
  gateways:
  - case-d-gateway-1
  http:
  - match:
    - uri:
        prefix: /fortio1/
    route:
    - destination:
        port:
          number: 80
        host: fortio-svc-1
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: case-d-virtualservice-2
  namespace: chenxma
spec:
  hosts:
  - "*"
  gateways:
  - case-d-gateway-2
  http:
  - match:
    - uri:
        prefix: /fortio2/
    route:
    - destination:
        port:
          number: 80
        host: fortio-svc-1
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: case-d-virtualservice-3
  namespace: chenxma
spec:
  hosts:
  - "*"
  gateways:
  - case-d-gateway-3
  http:
  - match:
    - uri:
        prefix: /fortio3/
    route:
    - destination:
        port:
          number: 80
        host: fortio-svc-1
```
and now if do:
```
curl <gateway_svc_ip>:80/fortio1/
curl <gateway_svc_ip>:81/fortio2/
curl <gateway_svc_ip>:82/fortio3/
```
I will get 200 return codes as expected.

But soon I found the following cases also return 200:
```
curl <ingressgateway_svc_ip>:80/fortio2/
curl <ingressgateway_svc_ip>:80/fortio3/
curl <ingressgateway_svc_ip>:81/fortio1/
curl <ingressgateway_svc_ip>:81/fortio3/
curl <ingressgateway_svc_ip>:82/fortio1/
curl <ingressgateway_svc_ip>:82/fortio2/
```
which I think is not expected according to Istio docs.
So I get into my Ingressgateway pod and dump the routing configs:
```
{
   "@type": "type.googleapis.com/envoy.admin.v2alpha.RoutesConfigDump",
   "dynamic_route_configs": [
    {
     "version_info": "2018-12-06T08:02:17Z",
     "route_config": {
      "name": "http.80",
      "virtual_hosts": [
       {
        "name": "*:80",
        "domains": [
         "*",
         "*:80"
        ],
        "routes": [
         {
          "match": {
           "prefix": "/fortio1/"
          },
          "route": {
           "cluster": "outbound|80||fortio-svc-1.chenxma.svc.cluster.local",
           "timeout": "0s",
           "max_grpc_timeout": "0s"
          },
          "decorator": {
           "operation": "fortio-svc-1.chenxma.svc.cluster.local:80/fortio1/*"
          },
          "per_filter_config": {
           "mixer": {
            "forward_attributes": {
             "attributes": {
              "destination.service.namespace": {
               "string_value": "chenxma"
              },
              "destination.service.name": {
               "string_value": "fortio-svc-1"
              },
              "destination.service": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.uid": {
               "string_value": "istio://chenxma/services/fortio-svc-1"
              },
              "destination.service.host": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              }
             }
            },
            "mixer_attributes": {
             "attributes": {
              "destination.service.uid": {
               "string_value": "istio://chenxma/services/fortio-svc-1"
              },
              "destination.service.host": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.namespace": {
               "string_value": "chenxma"
              },
              "destination.service.name": {
               "string_value": "fortio-svc-1"
              },
              "destination.service": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              }
             }
            }
           }
          }
         },
         {
          "match": {
           "prefix": "/fortio3/"
          },
          "route": {
           "cluster": "outbound|80||fortio-svc-1.chenxma.svc.cluster.local",
           "timeout": "0s",
           "max_grpc_timeout": "0s"
          },
          "decorator": {
           "operation": "fortio-svc-1.chenxma.svc.cluster.local:80/fortio3/*"
          },
          "per_filter_config": {
           "mixer": {
            "mixer_attributes": {
             "attributes": {
              "destination.service.host": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.uid": {
               "string_value": "istio://chenxma/services/fortio-svc-1"
              },
              "destination.service.name": {
               "string_value": "fortio-svc-1"
              },
              "destination.service.namespace": {
               "string_value": "chenxma"
              },
              "destination.service": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              }
             }
            },
            "forward_attributes": {
             "attributes": {
              "destination.service.name": {
               "string_value": "fortio-svc-1"
              },
              "destination.service.namespace": {
               "string_value": "chenxma"
              },
              "destination.service": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.host": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.uid": {
               "string_value": "istio://chenxma/services/fortio-svc-1"
              }
             }
            }
           }
          }
         },
         {
          "match": {
           "prefix": "/fortio2/"
          },
          "route": {
           "cluster": "outbound|80||fortio-svc-1.chenxma.svc.cluster.local",
           "timeout": "0s",
           "max_grpc_timeout": "0s"
          },
          "decorator": {
           "operation": "fortio-svc-1.chenxma.svc.cluster.local:80/fortio2/*"
          },
          "per_filter_config": {
           "mixer": {
            "mixer_attributes": {
             "attributes": {
              "destination.service.namespace": {
               "string_value": "chenxma"
              },
              "destination.service.name": {
               "string_value": "fortio-svc-1"
              },
              "destination.service": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.uid": {
               "string_value": "istio://chenxma/services/fortio-svc-1"
              },
              "destination.service.host": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              }
             }
            },
            "forward_attributes": {
             "attributes": {
              "destination.service.uid": {
               "string_value": "istio://chenxma/services/fortio-svc-1"
              },
              "destination.service.host": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.namespace": {
               "string_value": "chenxma"
              },
              "destination.service.name": {
               "string_value": "fortio-svc-1"
              },
              "destination.service": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              }
             }
            }
           }
          }
         }
        ]
       }
      ],
      "validate_clusters": false
     },
     "last_updated": "2018-12-06T08:02:44.814Z"
    },
    {
     "version_info": "2018-12-06T08:02:17Z",
     "route_config": {
      "name": "http.82",
      "virtual_hosts": [
       {
        "name": "*:82",
        "domains": [
         "*",
         "*:82"
        ],
        "routes": [
         {
          "match": {
           "prefix": "/fortio1/"
          },
          "route": {
           "cluster": "outbound|80||fortio-svc-1.chenxma.svc.cluster.local",
           "timeout": "0s",
           "max_grpc_timeout": "0s"
          },
          "decorator": {
           "operation": "fortio-svc-1.chenxma.svc.cluster.local:80/fortio1/*"
          },
          "per_filter_config": {
           "mixer": {
            "forward_attributes": {
             "attributes": {
              "destination.service.host": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.uid": {
               "string_value": "istio://chenxma/services/fortio-svc-1"
              },
              "destination.service.namespace": {
               "string_value": "chenxma"
              },
              "destination.service.name": {
               "string_value": "fortio-svc-1"
              },
              "destination.service": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              }
             }
            },
            "mixer_attributes": {
             "attributes": {
              "destination.service.name": {
               "string_value": "fortio-svc-1"
              },
              "destination.service.namespace": {
               "string_value": "chenxma"
              },
              "destination.service": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.host": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.uid": {
               "string_value": "istio://chenxma/services/fortio-svc-1"
              }
             }
            }
           }
          }
         },
         {
          "match": {
           "prefix": "/fortio3/"
          },
          "route": {
           "cluster": "outbound|80||fortio-svc-1.chenxma.svc.cluster.local",
           "timeout": "0s",
           "max_grpc_timeout": "0s"
          },
          "decorator": {
           "operation": "fortio-svc-1.chenxma.svc.cluster.local:80/fortio3/*"
          },
          "per_filter_config": {
           "mixer": {
            "forward_attributes": {
             "attributes": {
              "destination.service.namespace": {
               "string_value": "chenxma"
              },
              "destination.service.name": {
               "string_value": "fortio-svc-1"
              },
              "destination.service": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.host": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.uid": {
               "string_value": "istio://chenxma/services/fortio-svc-1"
              }
             }
            },
            "mixer_attributes": {
             "attributes": {
              "destination.service.namespace": {
               "string_value": "chenxma"
              },
              "destination.service.name": {
               "string_value": "fortio-svc-1"
              },
              "destination.service": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.uid": {
               "string_value": "istio://chenxma/services/fortio-svc-1"
              },
              "destination.service.host": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              }
             }
            }
           }
          }
         },
         {
          "match": {
           "prefix": "/fortio2/"
          },
          "route": {
           "cluster": "outbound|80||fortio-svc-1.chenxma.svc.cluster.local",
           "timeout": "0s",
           "max_grpc_timeout": "0s"
          },
          "decorator": {
           "operation": "fortio-svc-1.chenxma.svc.cluster.local:80/fortio2/*"
          },
          "per_filter_config": {
           "mixer": {
            "forward_attributes": {
             "attributes": {
              "destination.service.host": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.uid": {
               "string_value": "istio://chenxma/services/fortio-svc-1"
              },
              "destination.service.namespace": {
               "string_value": "chenxma"
              },
              "destination.service.name": {
               "string_value": "fortio-svc-1"
              },
              "destination.service": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              }
             }
            },
            "mixer_attributes": {
             "attributes": {
              "destination.service.uid": {
               "string_value": "istio://chenxma/services/fortio-svc-1"
              },
              "destination.service.host": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.namespace": {
               "string_value": "chenxma"
              },
              "destination.service.name": {
               "string_value": "fortio-svc-1"
              },
              "destination.service": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              }
             }
            }
           }
          }
         }
        ]
       }
      ],
      "validate_clusters": false
     },
     "last_updated": "2018-12-06T08:02:44.814Z"
    },
    {
     "version_info": "2018-12-06T08:02:17Z",
     "route_config": {
      "name": "http.81",
      "virtual_hosts": [
       {
        "name": "*:81",
        "domains": [
         "*",
         "*:81"
        ],
        "routes": [
         {
          "match": {
           "prefix": "/fortio1/"
          },
          "route": {
           "cluster": "outbound|80||fortio-svc-1.chenxma.svc.cluster.local",
           "timeout": "0s",
           "max_grpc_timeout": "0s"
          },
          "decorator": {
           "operation": "fortio-svc-1.chenxma.svc.cluster.local:80/fortio1/*"
          },
          "per_filter_config": {
           "mixer": {
            "forward_attributes": {
             "attributes": {
              "destination.service.namespace": {
               "string_value": "chenxma"
              },
              "destination.service.name": {
               "string_value": "fortio-svc-1"
              },
              "destination.service": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.uid": {
               "string_value": "istio://chenxma/services/fortio-svc-1"
              },
              "destination.service.host": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              }
             }
            },
            "mixer_attributes": {
             "attributes": {
              "destination.service.host": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.uid": {
               "string_value": "istio://chenxma/services/fortio-svc-1"
              },
              "destination.service.namespace": {
               "string_value": "chenxma"
              },
              "destination.service.name": {
               "string_value": "fortio-svc-1"
              },
              "destination.service": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              }
             }
            }
           }
          }
         },
         {
          "match": {
           "prefix": "/fortio3/"
          },
          "route": {
           "cluster": "outbound|80||fortio-svc-1.chenxma.svc.cluster.local",
           "timeout": "0s",
           "max_grpc_timeout": "0s"
          },
          "decorator": {
           "operation": "fortio-svc-1.chenxma.svc.cluster.local:80/fortio3/*"
          },
          "per_filter_config": {
           "mixer": {
            "forward_attributes": {
             "attributes": {
              "destination.service.namespace": {
               "string_value": "chenxma"
              },
              "destination.service.name": {
               "string_value": "fortio-svc-1"
              },
              "destination.service": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.uid": {
               "string_value": "istio://chenxma/services/fortio-svc-1"
              },
              "destination.service.host": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              }
             }
            },
            "mixer_attributes": {
             "attributes": {
              "destination.service.host": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.uid": {
               "string_value": "istio://chenxma/services/fortio-svc-1"
              },
              "destination.service.namespace": {
               "string_value": "chenxma"
              },
              "destination.service.name": {
               "string_value": "fortio-svc-1"
              },
              "destination.service": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              }
             }
            }
           }
          }
         },
         {
          "match": {
           "prefix": "/fortio2/"
          },
          "route": {
           "cluster": "outbound|80||fortio-svc-1.chenxma.svc.cluster.local",
           "timeout": "0s",
           "max_grpc_timeout": "0s"
          },
          "decorator": {
           "operation": "fortio-svc-1.chenxma.svc.cluster.local:80/fortio2/*"
          },
          "per_filter_config": {
           "mixer": {
            "forward_attributes": {
             "attributes": {
              "destination.service.namespace": {
               "string_value": "chenxma"
              },
              "destination.service.name": {
               "string_value": "fortio-svc-1"
              },
              "destination.service": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.uid": {
               "string_value": "istio://chenxma/services/fortio-svc-1"
              },
              "destination.service.host": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              }
             }
            },
            "mixer_attributes": {
             "attributes": {
              "destination.service.namespace": {
               "string_value": "chenxma"
              },
              "destination.service.name": {
               "string_value": "fortio-svc-1"
              },
              "destination.service": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.host": {
               "string_value": "fortio-svc-1.chenxma.svc.cluster.local"
              },
              "destination.service.uid": {
               "string_value": "istio://chenxma/services/fortio-svc-1"
              }
             }
            }
           }
          }
         }
        ]
       }
      ],
      "validate_clusters": false
     },
     "last_updated": "2018-12-06T08:02:44.814Z"
    }
   ]
  }
```
Note for each `route_config`, all the three VirtualServices rules(i.e. prefix `/fortio1/`, `/fortio2/` and `/fortio3/`) have been included. 
**Expected behavior**
With my VirtualService and Gateway setup, the following requests should return 404:
```
curl <ingressgateway_svc_ip>:80/fortio2/
curl <ingressgateway_svc_ip>:80/fortio3/
curl <ingressgateway_svc_ip>:81/fortio1/
curl <ingressgateway_svc_ip>:81/fortio3/
curl <ingressgateway_svc_ip>:82/fortio1/
curl <ingressgateway_svc_ip>:82/fortio2/
```

**Steps to reproduce the bug**
described above

**Version**
Kubernetes: 
```
Server Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.5-413+a6fbb4730262ed-dirty-dirty", GitCommit:"a6fbb4730262ed86a30aa1543631784f67be6692", GitTreeState:"dirty", BuildDate:"2018-12-05T10:14:34Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```
Istio: 1.0.2

**Installation**
manually installed via `istio-demo.yaml` with mesh disabled

**Environment**
CentOS Linux release 7.5.1804 (Core)

**Cluster state**
N/A

**Comment by arnecls** [comment]
That's an issue we also ran into.
Our current view on the problem is that the 3 gateway definitions are internally merged into one - which is _very_ confusing. It would be great if someone from the istio team could give a proper explanation what causes this / if this is expected behavior.

We fixed this by having only one gateway with all ports and add ports to the matching rules in the virtual services.

**Comment by edussx** [comment]
@arnecls Thanks for the information! At least I can confirm it's not a single problem on my side.

In my case, I need namespace separation so probably one large gateway including everything doesn't work :(

**Comment by linsun** [comment]
This issue was discussed briefly in today's istio user community call.  @GregHanson mentioned each port should have a unique port name  they are all currently called “http”.  

would be great to retry with unique port name.

**Comment by GregHanson** [comment]
A few things just looking at the config:
- all ports have the same name "http", each port should have a unique names such as `http-fortio1`, `http-fortio2`, etc. I know this is a requirement for apps/services and seems like you might get into trouble with duplicate names. Port naming conventions can be found here: https://istio.io/docs/setup/kubernetes/spec-requirements/  
- it is generally expected for there to be one resource per hostname.  In this case, there are 3 Gateway resources with the same host `"*"`.  There is some merge logic for Gateways so this is probably why it's happening

**Comment by GregHanson** [comment]
@amyroh ^^

**Comment by amyroh** [comment]
@GregHanson  @linsun Thanks for your input.  I think there is still an issue with merge logic for Gateways.  I tried the following Gateways with different hosts and http port names and still all these urls are accessible?  

http://test1.com:80/hello1
http://test1.com:81/hello2
http://test1.com:82/hello3
http://test2.com:80/hello1
http://test2.com:81/hello2
http://test2.com:82/hello3
http://test3.com:80/hello1
http://test3.com:81/hello2
http://test3.com:82/hello3

```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: helloworld-gateway-1
spec:
  selector:
    istio: ingressgateway # use istio default controller
  servers:
  - port:
      number: 80
      name: http1
      protocol: HTTP
    hosts:
    - "test1.com"
---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: helloworld-gateway-2
spec:
  selector:
    istio: ingressgateway # use istio default controller
  servers:
  - port:
      number: 81
      name: http2
      protocol: HTTP
    hosts:
    - "test2.com"
---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: helloworld-gateway-3
spec:
  selector:
    istio: ingressgateway # use istio default controller
  servers:
  - port:
      number: 82
      name: http3
      protocol: HTTP
    hosts:
    - "test3.com"
---

apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: helloworld-vs-1
spec:
  hosts:
  - "test1.com"
  gateways:
  - helloworld-gateway-1
  http:
  - match:
    - uri:
        exact: /hello1
    route:
    - destination:
        host: helloworld
        port:
          number: 5000
  - route:
    - destination:
        host: helloworld 
        port:
          number: 5000
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: helloworld-vs-2
spec:
  hosts:
  - "test2.com"
  gateways:
  - helloworld-gateway-2
  http:
  - match:
    - uri:
        exact: /hello2
    route:
    - destination:
        host: helloworld
        port:
          number: 5000
  - route:
    - destination:
        host: helloworld
        port:
          number: 5000
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: helloworld-vs-3
spec:
  hosts:
  - "test3.com"
  gateways:
  - helloworld-gateway-3
  http:
  - match:
    - uri:
        exact: /hello2
    route:
    - destination:
        host: helloworld
        port:
          number: 5000
  - route:
    - destination:
        host: helloworld
        port:
          number: 5000
```

**Comment by qichaoch** [comment]
@arnecls Thanks for the information. Your workaround fixes my problem.    I agree that there should be something wrong in the "merge" logic which cause the problem we met.   Changing the port name to be unique seems not working for this scenario.

**Comment by GregHanson** [comment]
@amyroh in your case, do you have 3 ingressgateways deployed with labels `helloworld-gateway-1`, `helloworld-gateway-2`, and `helloworld-gateway-3` respectively?  In your VirtualService specs:
```
  gateways:
  - helloworld-gateway-3
```
expects there to be a ingress gateway pod with `helloworld-gateway-3`

**Comment by edussx** [comment]
@GregHanson one gateway per hostname is not applicable in this case: https://github.com/istio/istio/issues/6071  

Basically TLS will stop working with `hosts` set to be anything not `*`. Please also see my comment: https://github.com/istio/istio/issues/6071#issuecomment-470520894

**Comment by rshriram** [comment]
fixed by #12663
